sql: fix RENAME COLUMN for REGIONAL BY ROW

otan opened the pull request:
We need to correct the column reference on the table descriptor if we
are renaming the REGIONAL BY ROW column.

Resolves #59116 

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60638)
<!-- Reviewable:end -->

ajstorm reviewed (changes requested):
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60638)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/rename_column.go, line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60638#-MThrN1i6ynXo41ykD_i:-MThrN1i6ynXo41ykD_j:b-naybfa) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bfd907bbb178f442d684133fb795094871c06e2/pkg/sql/rename_column.go#L266)):*
> ```Go
> 					pgerror.Newf(
> 						pgcode.InvalidColumnReference,
> 						"cannot rename column %s as it is was implicitly created as the REGIONAL BY ROW column reference",
> ```

Not sure I'm following the motivation for blocking here.  If we can get there by first changing to AS and then renaming, why not allow the rename directly here (perhaps with a warning that this will make the column visible?  Or maybe we leave it NOT VISIBLE but just change the name?).


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/rename_column.go, line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60638#-MThrN1i6ynXo41ykD_i:-MThvToYBWQ8gzaPbWon:b88smi5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bfd907bbb178f442d684133fb795094871c06e2/pkg/sql/rename_column.go#L266)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Not sure I'm following the motivation for blocking here.  If we can get there by first changing to AS and then renaming, why not allow the rename directly here (perhaps with a warning that this will make the column visible?  Or maybe we leave it NOT VISIBLE but just change the name?).
</blockquote></details>

My concern is keeping track that it was implicitly created. But if we don't care, I'll remove this check.


<!-- Sent from Reviewable.io -->

otan committed:
sql: fix RENAME COLUMN for REGIONAL BY ROW

We need to correct the column reference on the table descriptor if we
are renaming the REGIONAL BY ROW column.

Release note: None

ajstorm reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60638)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @otan)

___
*[pkg/sql/rename_column.go, line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60638#-MThrN1i6ynXo41ykD_i:-MThx5_c7l5TJzPHA039:b2a5sno) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bfd907bbb178f442d684133fb795094871c06e2/pkg/sql/rename_column.go#L266)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

My concern is keeping track that it was implicitly created. But if we don't care, I'll remove this check.
</blockquote></details>

Keeping track in what way?  Do we get a log entry of this rename?  Is that sufficient?  


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/rename_column.go, line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60638#-MThrN1i6ynXo41ykD_i:-MThyRWEENSaO-NtChUC:b3ce2ge) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bfd907bbb178f442d684133fb795094871c06e2/pkg/sql/rename_column.go#L266)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Keeping track in what way?  Do we get a log entry of this rename?  Is that sufficient?  
</blockquote></details>

Nah I meant if we decide to do something like eg drop the crdb region column when altering locality away from rbr. 


<!-- Sent from Reviewable.io -->

ajstorm reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60638#-:-MTk-mB68_IizkEdLI3q:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/rename_column.go, line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60638#-MThrN1i6ynXo41ykD_i:-MTjzpcs8ge1aiqgvTuI:b-c6hrmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bfd907bbb178f442d684133fb795094871c06e2/pkg/sql/rename_column.go#L266)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Nah I meant if we decide to do something like eg drop the crdb region column when altering locality away from rbr. 
</blockquote></details>

Ahh, in that case, I think it would be acceptable to not drop if the user explicitly renamed.  My take is that once they do the rename, they're taking ownership of this column.


<!-- Sent from Reviewable.io -->

otan commented:
bors r=ajstorm
